File: TP2/questao6.c
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <time.h> 
#include <locale.h>

// DEFINE DE VARIAVEIS

#define TAM_STRING         100 
#define TAM_STRING_ARRAY   100

typedef struct {
    int day;
    int month;
    int year;
} Date;

typedef struct{

    char id[TAM_STRING];
    char name[TAM_STRING];
    int countnames;
    char alternate_names[TAM_STRING_ARRAY][30];
    char house[TAM_STRING];
    char ancestry[TAM_STRING];
    char species[TAM_STRING];
    char patronus[TAM_STRING];
    bool hogwartsStaff;
    char hogwartsStudent[TAM_STRING];
    char actorName[TAM_STRING];
    bool alive;
    Date dateOfBirth;
    int yearOfBirth;
    char eyeColour[TAM_STRING];
    char gender[TAM_STRING];
    char hairColour[TAM_STRING];
    bool wizard;

}Personagem;

//-------------------------------------------
void id(char id[]);
void name(char name[]);
void alternate_names(char alternatename[]);
void house(char house[]);
void ancestry(char ancestry[]);
void species(char species[]);
void patronus(char patronus[]);
void hogwartsStaff(bool hogwartsStaff);
void hogwartsStudent(char hogwartsStudent[]);
void actorName(char actorName[]);
void alive(bool alive);
void dateOfBirth(Date dateOfBirth);
void yearOfBirth(char yearOfBirth[]);
void eyeColour(char eyeColour[]);
void gender(char gender[]);
void hairColour(char hairColour[]);
void wizard(bool wizard);
//--------------------------------------------
void swap(int i, int j);
void SelecaoRec(int i, int n);
void LerPersonagem(char entradaID[]);
void TratarString(char entrada[]);
Personagem clone(Personagem *personagem);
void imprimir();

Personagem lista[1000];
int contPersonagem = 0;
int contador = 0;

int main(){
    clock_t t; 
    t = clock(); 

    char Ids[1000];
    scanf("%s",Ids);

    while(strcmp(Ids,"FIM") != 0 ){      
        LerPersonagem(Ids);
        contPersonagem++;
        scanf("%s",Ids);
    }
    //Ordenacao do array de jogadores
    SelecaoRec(0 , contPersonagem);
    imprimir();

    // arquivo de matricula========================================================
    t = clock() - t;  
    FILE *arq;
    arq = fopen("matrícula_selecaoRecursiva.txt", "a");
    fprintf(arq, "695161 \t %ld \t %d", t , contador);
    fclose(arq);
    //=============================================================================

    return 0;
}

// GETS E SETS

void id(char id[]) {
    strcpy(lista[contPersonagem].id, id);
}
void name(char name[]) {
    strcpy(lista[contPersonagem].name, name);
}
void alternate_names(char alternatename[]) {
    char* nomeAlt = strtok(alternatename, ",");
    int count = 0;
    while (nomeAlt != NULL && count < TAM_STRING_ARRAY) {
        strcpy(lista[contPersonagem].alternate_names[count], nomeAlt);
        count++;
        nomeAlt = strtok(NULL, ",");
    }
    lista[contPersonagem].countnames = count;
}
void house(char house[]) {
    strcpy(lista[contPersonagem].house, house);
}
void ancestry(char ancestry[]) {
    strcpy(lista[contPersonagem].ancestry, ancestry);
}
void species(char species[]) {
    strcpy(lista[contPersonagem].species, species);
}
void patronus(char patronus[]) {
    strcpy(lista[contPersonagem].patronus, patronus);
}
void hogwartsStaff(bool hogwartsStaff) {
    lista[contPersonagem].hogwartsStaff = hogwartsStaff;
}
void hogwartsStudent(char hogwartsStudent[]) {
    strcpy(lista[contPersonagem].hogwartsStudent, hogwartsStudent);
}
void actorName(char actorName[]) {
    strcpy(lista[contPersonagem].actorName, actorName);
}
void alive(bool alive) {
    lista[contPersonagem].alive = alive;
}
void dateOfBirth(Date dateOfBirth) {
    lista[contPersonagem].dateOfBirth = dateOfBirth;
}
void yearOfBirth(char yearOfBirth[]) {
    lista[contPersonagem].yearOfBirth = atoi(yearOfBirth);
}
void eyeColour(char eyeColour[]) {
    strcpy(lista[contPersonagem].eyeColour, eyeColour);
}
void gender(char gender[]) {
    strcpy(lista[contPersonagem].gender, gender);
}
void hairColour(char hairColour[]) {
    strcpy(lista[contPersonagem].hairColour, hairColour);
}
void wizard(bool wizard) {
    lista[contPersonagem].wizard = wizard;
}

// ORDENACAO PARA PESQUISA BINARIA

// Selection sort 

void SelecaoRec(int i, int n){
	int menor = i;
	for (int j = i + 1; j < n; j++){
		if (strcmp(lista[j].name,lista[menor].name) < 0 ){
			menor = j;
            contador++;	
        }
	}
	swap(menor, i);

	if (i + 1 < n) {
		SelecaoRec(i + 1, n);
        contador++;
	}
}

void swap(int i, int j){
   Personagem tmp = lista[i];
   lista[i] = lista[j];
   lista[j] = tmp;
}

// LEITURA

void LerPersonagem(char entradaID[]){

    char entradas[1000];
    char *stringsep;
    char *virgula;
    FILE *caminho = fopen("/tmp/characters.csv","r");
    //C:\\Users\\WazX\\Desktop\\aeds2-master\\tps\\entrada e saida\\tp02\\characters.csv    /tmp/characters.csv

    do{
        fgets(entradas,1000,caminho);
        stringsep = strdup(entradas);
        virgula = strtok(stringsep,";");
        if (strcmp(virgula,entradaID) == 0 ){
            TratarString(entradas);
        }  
    }while(!feof(caminho) && strcmp(virgula,entradaID) != 0);
    fclose(caminho);

}

void TratarString(char entrada[]){
    //ID
    strcpy(entrada, strtok(entrada, ";")); 
    id(entrada);    
    //Name
    strcpy(entrada, strtok(NULL, ";"));
    name(entrada);
    //Alternate_Names
    //strcpy(entrada, strtok(NULL, ";"));
    //Alternate_Names(entrada);  
    //house
    strcpy(entrada, strtok(NULL, ";"));
    house(entrada);
    //patronous
    strcpy(entrada, strtok(NULL, ";"));
    patronus(entrada);   
    //hogwartsStaff
    strcpy(entrada, strtok(NULL, ";"));
    hogwartsStaff(entrada);
    //hogwartsStudent
    strcpy(entrada, strtok(NULL, ";"));
    hogwartsStudent(entrada);
    //actorName
    strcpy(entrada, strtok(NULL, ";"));
    actorName(entrada);
    //alive
    strcpy(entrada, strtok(NULL, ";"));
    alive(entrada);
    //dateOfBirth
    Date data;
    sscanf(entrada, "%d-%d-%d", &data.year, &data.month, &data.day);
    dateOfBirth(data);
    //yearOfBirth
    strcpy(entrada, strtok(NULL, ";"));
    yearOfBirth(entrada);
    //eyeColour
    strcpy(entrada, strtok(NULL, ";"));
    eyeColour(entrada);
    //gender
    strcpy(entrada, strtok(NULL, ";"));
    gender(entrada);
    //hairColour
    strcpy(entrada, strtok(NULL, ";"));
    hairColour(entrada);
    //wizard
    strcpy(entrada, strtok(NULL, ";"));
    wizard(entrada);
}

// CLONE

Personagem clone(Personagem *personagem){

    Personagem clone;

    strcpy(clone.id , personagem -> id);
    strcpy(clone.name , personagem -> name);
    for(int i = 0; i < personagem -> countnames; i++){
        strcpy(clone.alternate_names[i], personagem -> alternate_names[i]);
    } 
    strcpy(clone.house , personagem -> house);
    strcpy(clone.ancestry , personagem -> ancestry);
    strcpy(clone.species , personagem -> species);
    strcpy(clone.patronus , personagem -> patronus);
    clone.hogwartsStaff = personagem -> hogwartsStaff;
    strcpy(clone.hogwartsStudent , personagem -> hogwartsStudent);
    strcpy(clone.actorName , personagem -> actorName);
    clone.alive = personagem -> alive;
    clone.dateOfBirth = personagem -> dateOfBirth ;
    clone.yearOfBirth = personagem -> yearOfBirth ;
    strcpy(clone.eyeColour , personagem -> eyeColour);
    strcpy(clone.gender , personagem -> gender);
    strcpy(clone.hairColour , personagem -> hairColour);

    return clone;
}
// IMPRIMIR

void imprimir(){
    for (int i = 0; i < contPersonagem; i++) {
        printf("[%s", lista[i].id);
        printf("%s"," ## ");
        printf("%s", lista[i].name);
        printf("%s"," ## {");
        for (int j = 0; j < lista[i].countnames; j++) {
            printf("%s%s", lista[i].alternate_names[j], (j < lista[i].countnames - 1) ? ", " : "");
        }
        printf("%s", lista[i].house);
        printf("%s", lista[i].ancestry);
        printf("%s", lista[i].species);
        printf("%s", lista[i].patronus);
        printf("%s"," ## ");
        printf("%s", lista[i].hogwartsStaff ? "true" : "false");
        printf("%s"," ## ");
        printf("%s", lista[i].hogwartsStudent);
        printf("%s"," ## ");
        printf("%s", lista[i].actorName);
        printf("%s"," ## ");
        printf("%02d-%02d-%02d", lista[i].dateOfBirth.day, lista[i].dateOfBirth.month, lista[i].dateOfBirth.year);
        printf("%s"," ## ");
        printf("%d", lista[i].yearOfBirth);
        printf("%s"," ## ");
        printf("%s", lista[i].eyeColour);
        printf("%s"," ## ");
        printf("%s", lista[i].gender);
        printf("%s"," ## ");
        printf("%s", lista[i].hairColour);
        printf("%s"," ## ");
        printf("%s", lista[i].wizard ? "true" : "false");
        printf("%s","]\n");
    }
}

File: TP4/questao6.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;

class Personagem{
    public static Personagem[] personagens = new Personagem[1000];
    public static int tampersonagens = 0;

    private String id;
    private String name;
    private String[] alternate_names;
    private String house;
    private String ancestry;
    private String species;
    private String patronus;
    private Boolean hogwartsStaff;
    private Boolean hogwartsStudent;
    private String actorName;
    private Boolean alive;
    private Date dateOfBirth;
    private int yearOfBirth;
    private String eyeColour;
    private String gender;
    private String hairColour;
    private Boolean wizard;

    // CONSTRUTORES

    public Personagem(){
        id = "";
        name = "";
        alternate_names = new String[1];
        house = "";
        ancestry = "";
        species = "";
        patronus = "";
        hogwartsStaff = false;
        hogwartsStudent = false;
        actorName = "";
        alive = false;
        dateOfBirth = new Date(0);
        yearOfBirth = 0;
        eyeColour = "";
        gender = "";
        hairColour = "";
        wizard = false;
    }

    public Personagem(String id,String name,String[] alternate_names,String house,String ancestry,String species,String patronus,
    Boolean hogwartsStaff,Boolean hogwartsStudent,String actorName,Boolean alive,Date dateOfBirth,int yearOfBirth,String eyeColour,
    String gender,String hairColour,Boolean wizard){
        
        this.id = id ;
        this.name = name;
        this.alternate_names =alternate_names;
        this.house = house;
        this.ancestry = ancestry;
        this.species = species;
        this.patronus = patronus;
        this.hogwartsStaff = hogwartsStaff;
        this.hogwartsStudent = hogwartsStudent;
        this.actorName = actorName;
        this.alive = alive;
        this.dateOfBirth = dateOfBirth;
        this.yearOfBirth = yearOfBirth;
        this.eyeColour = eyeColour;
        this.gender = gender;
        this.hairColour = hairColour;
        this.wizard = wizard;
        
    }

    // GETS E SETS
    public String getId() {
        return id;
    }
    public String getName() {
        return name;
    }
    public String[] getAlternate_names() {
        return alternate_names;
    }
    public String getHouse() {
        return house;
    }
    public String getAncestry() {
        return ancestry;
    }
    public String getSpecies() {
        return species;
    }
    public String getPatronus() {
        return patronus;
    }
    public Boolean getHogwartsStaff() {
        return hogwartsStaff;
    }
    public Boolean getHogwartsStudent() {
        return hogwartsStudent;
    }
    public String getActorName() {
        return actorName;
    }
    public Boolean getAlive() {
        return alive;
    }
    public Date getDateOfBirth() {
        return dateOfBirth;
    }
    public int getYearOfBirth() {
        return yearOfBirth;
    }
    public String getEyeColour() {
        return eyeColour;
    }
    public String getGender() {
        return gender;
    }
    public String getHairColour() {
        return hairColour;
    }
    public Boolean getWizard() {
        return wizard;
    }

    public void setId(String id) {
        this.id = id;
    }
    public void setName(String name) {
        this.name = name;
    }
    public void setAlternate_names(String[] alternate_names) {
        this.alternate_names = alternate_names;
    }
    public void setHouse(String house) {
        this.house = house;
    }
    public void setAncestry(String ancestry) {
        this.ancestry = ancestry;
    }
    public void setSpecies(String species) {
        this.species = species;
    }
    public void setPatronus(String patronus) {
        this.patronus = patronus;
    }
    public void setHogwartsStaff(Boolean hogwartsStaff) {
        this.hogwartsStaff = hogwartsStaff;
    }
    public void setHogwartsStudent(Boolean hogwartsStudent) {
        this.hogwartsStudent = hogwartsStudent;
    }
    public void setActorName(String actorName) {
        this.actorName = actorName;
    }
    public void setAlive(Boolean alive) {
        this.alive = alive;
    }
    public void setDateOfBirth(Date dateOfBirth) {
        this.dateOfBirth = dateOfBirth;
    }
    public void setYearOfBirth(int yearOfBirth) {
        this.yearOfBirth = yearOfBirth;
    }
    public void setEyeColour(String eyeColour) {
        this.eyeColour = eyeColour;
    }
    public void setGender(String gender) {
        this.gender = gender;
    }
    public void setHairColour(String hairColour) {
        this.hairColour = hairColour;
    }
    public void setWizard(Boolean wizard) {
        this.wizard = wizard;
    }

    // CLONE

    public Personagem Clone(){
        Personagem clone = new Personagem();

        clone.id = id ;
        clone.name = name;
        clone.alternate_names =alternate_names;
        clone.house = house;
        clone.ancestry = ancestry;
        clone.species = species;
        clone.patronus = patronus;
        clone.hogwartsStaff = hogwartsStaff;
        clone.hogwartsStudent = hogwartsStudent;
        clone.actorName = actorName;
        clone.alive = alive;
        clone.dateOfBirth = dateOfBirth;
        clone.yearOfBirth = yearOfBirth;
        clone.eyeColour = eyeColour;
        clone.gender = gender;
        clone.hairColour = hairColour;
        clone.wizard = wizard;

        return clone;
    }

    // IMPRIMIR
    public void imprimir(){
        SimpleDateFormat sdf = new SimpleDateFormat("dd-MM-yyyy");

        System.out.print(getId()+" ## "+getName()+" ## ");
        String[] alternateNames = getAlternate_names();
        System.out.print("{");
        for (int j = 0; j < alternateNames.length; j++) {
            System.out.print(alternateNames[j]);
            if (j < alternateNames.length - 1) {
                System.out.print(",");
            }
        }
        System.out.print("} ## "+getHouse()+
        " ## "+ getAncestry()+" ## "+ getSpecies()+" ## "+getPatronus()+
        " ## " +getHogwartsStaff()+" ## " +getHogwartsStudent()+" ## " +getActorName()+
        " ## " +getAlive()+" ## " + sdf.format(getDateOfBirth())+" ## " +getYearOfBirth()+
        " ## " +getEyeColour()+" ## " +getGender()+" ## " +getHairColour()+
        " ## " +getWizard()+"]\n");
    }
}

class HashComRehash {
	Personagem tabela[];
	int m;
	final int NULO = -1;

	public HashComRehash() {
		this(21);
	}

	public HashComRehash(int m) {
		this.m = m;
		this.tabela = new Personagem[this.m];
		for (int i = 0; i < m; i++) {
			tabela[i] = null;
		}
	}

	public int h(int elemento) {
		return elemento % m;
	}

	public int reh(int elemento) {
		return (elemento+1)% m;
	}

	public boolean inserir(Personagem elemento) {
		boolean resp = false;
        int soma = 0;
        for(int i = 0; i < elemento.getName().length(); i++){
            soma = soma + (int) elemento.getName().charAt(i);
        }
		if (elemento != null) {
			int pos = h(soma);
			if (tabela[pos] == null) {
				tabela[pos] = elemento;
				resp = true;
			} else {
				pos = reh(soma);
				if (tabela[pos] == null) {
				tabela[pos] = elemento;
				resp = true;
				}
			}
		}
		return resp;
	}

    public boolean pesquisar(int altura) {
        boolean resp = false;
        int pos = h(altura);

        // Procura na posição calculada pela função de hash
        if (tabela[pos] != null && tabela[pos].getYearOfBirth() == altura) {
            resp = true;
        } else {
            // Procura na posição recalculada pela função de rehash
            pos = reh(altura);
            if (tabela[pos] != null && tabela[pos].getYearOfBirth() == altura) {
                resp = true;
            }
        }
        return resp;
    }

	public boolean pesquisar(int elemento , String nomejog) {
        boolean resp = false;
        int pos = h(elemento);
        if (tabela[pos].getName().compareTo(nomejog) == 0) {
            resp = true;
        }else if (tabela[pos] != null) {
            pos = reh(elemento);
            if (tabela[pos].getName().compareTo(nomejog) == 0) {
                resp = true;
            }
        }
        return resp;
   }

    public boolean pesquisar(String elemento) {
        boolean resp = false;
        int soma = 0;
        for(int i = 0; i < elemento.length(); i++){
            soma = soma + (int) elemento.charAt(i);
        }
        int pos = h(soma);
        if (tabela[pos].getName().compareTo(elemento) == 0) {
            resp = true;
            questao6.contpos = pos;
        } else if (tabela[pos] != null) {
        pos = reh(soma);
        if (tabela[pos].getName().compareTo(elemento) == 0) {
            resp = true;
            questao6.contpos = pos;
        }
        }
        return resp;
    }
}

public class questao6{

    public static Personagem[] personagens = new Personagem[1000];
    public static int tampersonagens = 0;
    public static int contador = 0;
    public static int contpos = 0;

    // LEITURA

    public static String ler(String entradaid) throws Exception {
        String entrada = "";
        BufferedReader arq = new BufferedReader(new InputStreamReader(new FileInputStream("/tmp/characters.csv")));
        //C:\\Users\\WazX\\Desktop\\aeds2-master\\tps\\entrada e saida\\tp02\\characters.csv    /tmp/characters.csv

        entrada = arq.readLine();
        while(entrada != null){
            String[] Ids = entrada.split(";");
            if(Ids[0].equals(entradaid) == true){
                return entrada;
            }
            entrada = arq.readLine();
        }
        String saida = "";

        return saida;
    }

    public static void TratarString(String entrada)throws Exception{
        int tam = entrada.length();
        if (entrada.charAt(tam - 1) == ';') {
            entrada = entrada + "";
        }
        entrada = entrada.replaceAll(";;", ";;");
        String[] HPpersonagens = entrada.split(";");

        // ID do Personagem
        personagens[tampersonagens].setId(HPpersonagens[0]);
        // Nome do Personagem
        personagens[tampersonagens].setName(HPpersonagens[1]);
        // Alternate_names do Personagem
        int inicio = HPpersonagens[2].indexOf('[');
        int fim = HPpersonagens[2].indexOf(']');
        if (inicio != -1 && fim != -1 && fim > inicio) {
            String alternateNames = HPpersonagens[2].substring(inicio + 1, fim);
            alternateNames = alternateNames.replaceAll("'", "");
            personagens[tampersonagens].setAlternate_names(alternateNames.split(","));
        }
        // House do Personagem
        if (!HPpersonagens[3].equals("[]")) {
            personagens[tampersonagens].setHouse(HPpersonagens[3]);
        } else {
            personagens[tampersonagens].setHouse("");
        }
        // Ancestry do Personagem
        if (!HPpersonagens[4].equals("[]")) {
            personagens[tampersonagens].setAncestry(HPpersonagens[4]);
        } else {
            personagens[tampersonagens].setAncestry("");
        }
        // Species do Personagem
        if (!HPpersonagens[5].equals("[]")) {
            personagens[tampersonagens].setSpecies(HPpersonagens[5]);
        } else {
            personagens[tampersonagens].setSpecies("");
        }
        // Patronous do Personagem
        if (!HPpersonagens[6].equals("[]")) {
            personagens[tampersonagens].setPatronus(HPpersonagens[6]);
        } else {
            personagens[tampersonagens].setPatronus("");
        }
        // HogwartsStaff do Personagem
        boolean hogwartsStaff = Boolean.parseBoolean(HPpersonagens[7]);
        personagens[tampersonagens].setHogwartsStaff(hogwartsStaff);
        // Hogwartstudent do Personagem
        boolean hogwartsStudent = Boolean.parseBoolean(HPpersonagens[8]);
        personagens[tampersonagens].setHogwartsStudent(hogwartsStudent);
        // Ator do Personagem
        personagens[tampersonagens].setActorName(HPpersonagens[9]);
        // Personagem Alive
        boolean alive = Boolean.parseBoolean(HPpersonagens[10]);
        personagens[tampersonagens].setAlive(alive);
        // Date of birth do Personagem
        SimpleDateFormat sdf = new SimpleDateFormat("dd-MM-yyyy");
        if (!HPpersonagens[12].equals("[]")) {
            try {
                Date dataNascimento = sdf.parse(HPpersonagens[12]);
                personagens[tampersonagens].setDateOfBirth(dataNascimento);
            } catch (ParseException e) {
                personagens[tampersonagens].setDateOfBirth(new Date());
            }
        } else {
            personagens[tampersonagens].setDateOfBirth(new Date());
        }
        // Year of birth do Personagem
        int anonascimento = Integer.parseInt(HPpersonagens[13]);
        personagens[tampersonagens].setYearOfBirth(anonascimento);
        // Eye Colour do Personagem
        if (!HPpersonagens[14].equals("[]")) {
            personagens[tampersonagens].setEyeColour(HPpersonagens[14]);
        } else {
            personagens[tampersonagens].setEyeColour("");
        }
        // Gender do Personagem
        if (!HPpersonagens[15].equals("[]")) {
            personagens[tampersonagens].setGender(HPpersonagens[15]);
        } else {
            personagens[tampersonagens].setGender("");
        }
        // Hair Colour do Personagem
        if (!HPpersonagens[16].equals("[]")) {
            personagens[tampersonagens].setHairColour(HPpersonagens[16]);
        } else {
            personagens[tampersonagens].setHairColour("");
        }
        // Wizard do Personagem
        boolean wizard = HPpersonagens[17].equalsIgnoreCase("verdadeiro");
        personagens[tampersonagens].setWizard(wizard);
    }

    public static void main(String[] args) throws Exception {

        //tempo inicial do código
        long tempoInicial = System.currentTimeMillis();
        //----------------------------------------------//

        HashComRehash tabelaRehash = new HashComRehash();
        String IdsPersonagens = "";
        BufferedReader entrada = new BufferedReader(new InputStreamReader(System.in));
        IdsPersonagens = entrada.readLine();
        while(IdsPersonagens.equals("FIM") != true){
            personagens[tampersonagens] = new Personagem();
            TratarString(ler(IdsPersonagens));
            tampersonagens++;
            IdsPersonagens = entrada.readLine();
        }
        for(int i =0 ; i< tampersonagens ; i++){
            tabelaRehash.inserir(personagens[i]);
        }
        
        // SEGUNDA PARTE DA LEITURA
        String NomePersonagem = "";
        boolean saida = false;
        NomePersonagem = entrada.readLine();

        while(NomePersonagem.equals("FIM") != true){
            saida = tabelaRehash.pesquisar(NomePersonagem);
            if(saida == false){
                System.out.println(NomePersonagem + " NAO");
            }else{
                System.out.println(NomePersonagem + " (pos: "+ contpos +") SIM"); 
            }
            NomePersonagem = entrada.readLine();
        }

        //arquivo de Matricula sequencial 
        long tempoFinal = System.currentTimeMillis();
        Arq.openWrite("matrícula_hashRehash.txt");
        Arq.println("695161" + "\t" + (tempoFinal - tempoInicial) + "\t");
        Arq.close();
    }
}
